### ISSUE TITLE
Some problems in ODE results

### ISSUE THREAD

**Issue by kquine** [dreal/dreal3#73]
Here is another example with linear ODEs, and there are problems in the result. I guess that it is cause by different variable orderings of define-ODE and integral. The drh file is as follows: 

```
#define g           9.80555
#define pi          3.1415926535897931
#define vT          92.827721
#define yBTA        -0.099593
#define lBTA        -1.700982
#define lP          -1.184647
#define lR           0.223908
#define nBTA         0.407420
#define nP          -0.056276
#define nR          -0.188010
#define yRDR         0.740361
#define lAIL         0.531304
#define lRDR         0.049766
#define nAIL         0.005685
#define nRDR        -0.106592

[-pi,pi]    beta;
[-pi,pi]    p;
[-pi,pi]    r;
[-pi,pi]    phi;
[-pi,pi]    psi;
[-pi,pi]    xAIL;
[-pi,pi]    rAIL;
[-pi,pi]    xRDR;
[-pi,pi]    rRDR;
[0,1]       tau;
[0,1]       time;

{mode 1;
  invt: 
  flow: 
	d/dt[beta]    = yBTA * beta - r + (g / vT) * phi + yRDR * xRDR;
	d/dt[p]       = lBTA * beta + lP * p + lR * r + lAIL * xAIL + lRDR * xRDR;
	d/dt[r]       = nBTA * beta + nP * p + nR * r + nAIL * xAIL + nRDR * xRDR;
	d/dt[phi]     = p;
	d/dt[psi]     = (g / vT) * phi;
	d/dt[xAIL]    = rAIL;
	d/dt[rAIL]    = 0;
	d/dt[xRDR]    = rRDR;
	d/dt[rRDR]    = 0;
	d/dt[tau]     = 1;
  jump: 
        (and (tau = 1)) ==> 
        @1 (and (beta' = beta) (p' = p) (r' = r) (phi' = phi) (psi' = psi) 
                (xAIL' = xAIL) (rAIL' = rAIL) 
                (xRDR' = xRDR) (rRDR' = rRDR) 
                (tau' = 0));
}

init: @1 (and (beta = 0)   (p = 0)       (r = 0)      (phi = 0)    (psi = 0)  
              (xAIL = 0)   (rAIL = 0.05) 
              (xRDR = 0)   (rRDR = 0.1)  
              (tau = 0));
goal: @1 true;
```

Below is the result given by `/dReach -k 5 -l 5 test-air-single.drh`. The variable rAIL changes and xRDR does not change, though `d/dt[rAIL] = 0`, `d/dt[xRDR] = rRDR` and `rRDR=0.1`:

```
SMT: test-air-single_5_0.smt2, PATH : [1,1,1,1,1,1]
Solution:
beta_0_0 : [-3.14159, 3.141590000000001] = [-0, 0]
beta_0_t : [-3.14159, 3.141590000000001] = [-0.03964001440156357, -0.03964001440156351]
beta_1_0 : [-3.14159, 3.141590000000001] = [-0.03964001440156357, -0.03964001440156351]
beta_1_t : [-3.14159, 3.141590000000001] = [-0.04817418460016381, -0.04817418460016362]
beta_2_0 : [-3.14159, 3.141590000000001] = [-0.04817418460016381, -0.04817418460016362]
beta_2_t : [-3.14159, 3.141590000000001] = [-0.02112500456768204, -0.02112500456768169]
beta_3_0 : [-3.14159, 3.141590000000001] = [-0.02112500456768204, -0.02112500456768169]
beta_3_t : [-3.14159, 3.141590000000001] = [0.02922670667795814, 0.0292267066779588]
beta_4_0 : [-3.14159, 3.141590000000001] = [0.02922670667795814, 0.0292267066779588]
beta_4_t : [-3.14159, 3.141590000000001] = [0.08091456382232736, 0.08091456382232862]
beta_5_0 : [-3.14159, 3.141590000000001] = [0.08091456382232736, 0.08091456382232862]
beta_5_t : [-3.14159, 3.141590000000001] = [0.08091456382232735, 0.08094595937594977]
mode_0 : [1, 1] = [1, 1]
mode_1 : [1, 1] = [1, 1]
mode_2 : [1, 1] = [1, 1]
mode_3 : [1, 1] = [1, 1]
mode_4 : [1, 1] = [1, 1]
mode_5 : [1, 1] = [1, 1]
p_0_0 : [-3.14159, 3.141590000000001] = [-0, 0]
p_0_t : [-3.14159, 3.141590000000001] = [0.04943732599400113, 0.04943732599400126]
p_1_0 : [-3.14159, 3.141590000000001] = [0.04943732599400113, 0.04943732599400126]
p_1_t : [-3.14159, 3.141590000000001] = [0.1145494190147177, 0.1145494190147181]
p_2_0 : [-3.14159, 3.141590000000001] = [0.1145494190147177, 0.1145494190147181]
p_2_t : [-3.14159, 3.141590000000001] = [0.1494070828275359, 0.1494070828275366]
p_3_0 : [-3.14159, 3.141590000000001] = [0.1494070828275359, 0.1494070828275366]
p_3_t : [-3.14159, 3.141590000000001] = [0.1474928476085329, 0.1474928476085341]
p_4_0 : [-3.14159, 3.141590000000001] = [0.1474928476085329, 0.1474928476085341]
p_4_t : [-3.14159, 3.141590000000001] = [0.1266212129548316, 0.1266212129548335]
p_5_0 : [-3.14159, 3.141590000000001] = [0.1266212129548316, 0.1266212129548335]
p_5_t : [-3.14159, 3.141590000000001] = [0.1266062949554142, 0.1266212129548335]
phi_0_0 : [-3.14159, 3.141590000000001] = [-0, 0]
phi_0_t : [-3.14159, 3.141590000000001] = [0.02002336871528148, 0.02002336871528154]
phi_1_0 : [-3.14159, 3.141590000000001] = [0.02002336871528148, 0.02002336871528154]
phi_1_t : [-3.14159, 3.141590000000001] = [0.1033674204716396, 0.10336742047164]
phi_2_0 : [-3.14159, 3.141590000000001] = [0.1033674204716396, 0.10336742047164]
phi_2_t : [-3.14159, 3.141590000000001] = [0.2386167881621738, 0.2386167881621749]
phi_3_0 : [-3.14159, 3.141590000000001] = [0.2386167881621738, 0.2386167881621749]
phi_3_t : [-3.14159, 3.141590000000001] = [0.3896492625257834, 0.3896492625257857]
phi_4_0 : [-3.14159, 3.141590000000001] = [0.3896492625257834, 0.3896492625257857]
phi_4_t : [-3.14159, 3.141590000000001] = [0.5271863701113847, 0.5271863701113893]
phi_5_0 : [-3.14159, 3.141590000000001] = [0.5271863701113847, 0.5271863701113893]
phi_5_t : [-3.14159, 3.141590000000001] = [0.5271863701113846, 0.5272746152592981]
psi_0_0 : [-3.14159, 3.141590000000001] = [-0, 0]
psi_0_t : [-3.14159, 3.141590000000001] = [0.0006062378829492108, 0.0006062378829492125]
psi_1_0 : [-3.14159, 3.141590000000001] = [0.0006062378829492108, 0.0006062378829492125]
psi_1_t : [-3.14159, 3.141590000000001] = [0.006543743348877986, 0.006543743348878011]
psi_2_0 : [-3.14159, 3.141590000000001] = [0.006543743348877986, 0.006543743348878011]
psi_2_t : [-3.14159, 3.141590000000001] = [0.02429833842660997, 0.02429833842661008]
psi_3_0 : [-3.14159, 3.141590000000001] = [0.02429833842660997, 0.02429833842661008]
psi_3_t : [-3.14159, 3.141590000000001] = [0.05750046904475369, 0.05750046904475403]
psi_4_0 : [-3.14159, 3.141590000000001] = [0.05750046904475369, 0.05750046904475403]
psi_4_t : [-3.14159, 3.141590000000001] = [0.1061118052748433, 0.1061118052748441]
psi_5_0 : [-3.14159, 3.141590000000001] = [0.1061118052748433, 0.1061118052748441]
psi_5_t : [-3.14159, 3.141590000000001] = [0.1061118052748433, 0.1061506204899354]
rAIL_0_0 : [-3.14159, 3.141590000000001] = [0.05000000000000001, 0.05000000000000001]
rAIL_0_t : [-3.14159, 3.141590000000001] = [0.03229734116496337, 0.03229734116496344]
rAIL_1_0 : [-3.14159, 3.141590000000001] = [0.03229734116496337, 0.03229734116496344]
rAIL_1_t : [-3.14159, 3.141590000000001] = [0.005731093024401153, 0.005731093024401265]
rAIL_2_0 : [-3.14159, 3.141590000000001] = [0.005731093024401153, 0.005731093024401265]
rAIL_2_t : [-3.14159, 3.141590000000001] = [-0.0146209151026667, -0.01462091510266649]
rAIL_3_0 : [-3.14159, 3.141590000000001] = [-0.0146209151026667, -0.01462091510266649]
rAIL_3_t : [-3.14159, 3.141590000000001] = [-0.01664271863152369, -0.01664271863152332]
rAIL_4_0 : [-3.14159, 3.141590000000001] = [-0.01664271863152369, -0.01664271863152332]
rAIL_4_t : [-3.14159, 3.141590000000001] = [0.002576600154985414, 0.002576600154986033]
rAIL_5_0 : [-3.14159, 3.141590000000001] = [0.002576600154985414, 0.002576600154986033]
rAIL_5_t : [-3.14159, 3.141590000000001] = [0.002576600154985413, 0.002596256982908554]
rRDR_0_0 : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_0_t : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_1_0 : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_1_t : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_2_0 : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_2_t : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_3_0 : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_3_t : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_4_0 : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_4_t : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_5_0 : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
rRDR_5_t : [-3.14159, 3.141590000000001] = [0.1000000000000001, 0.1000000000000001]
r_0_0 : [-3.14159, 3.141590000000001] = [-0, 0]
r_0_t : [-3.14159, 3.141590000000001] = [-0, 0]
r_1_0 : [-3.14159, 3.141590000000001] = [-0, 0]
r_1_t : [-3.14159, 3.141590000000001] = [-0, 0]
r_2_0 : [-3.14159, 3.141590000000001] = [-0, 0]
r_2_t : [-3.14159, 3.141590000000001] = [-0, 0]
r_3_0 : [-3.14159, 3.141590000000001] = [-0, 0]
r_3_t : [-3.14159, 3.141590000000001] = [-0, 0]
r_4_0 : [-3.14159, 3.141590000000001] = [-0, 0]
r_4_t : [-3.14159, 3.141590000000001] = [-0, 0]
r_5_0 : [-3.14159, 3.141590000000001] = [-0, 0]
r_5_t : [-3.14159, 3.141590000000001] = [-0, 0]
tau_0_0 : [0, 1] = [0, 0]
tau_0_t : [0, 1] = [1, 1]
tau_1_0 : [0, 1] = [0, 0]
tau_1_t : [0, 1] = [1, 1]
tau_2_0 : [0, 1] = [0, 0]
tau_2_t : [0, 1] = [1, 1]
tau_3_0 : [0, 1] = [0, 0]
tau_3_t : [0, 1] = [1, 1]
tau_4_0 : [0, 1] = [0, 0]
tau_4_t : [0, 1] = [1, 1]
tau_5_0 : [0, 1] = [0, 0]
tau_5_t : [0, 1] = [0, 0.0006195189361513088]
time_0 : [0, 1] = [0.9999999999999999, 1]
time_1 : [0, 1] = [0.9999999999999999, 1]
time_2 : [0, 1] = [0.9999999999999999, 1]
time_3 : [0, 1] = [0.9999999999999999, 1]
time_4 : [0, 1] = [0.9999999999999999, 1]
time_5 : [0, 1] = [0, 0.0006533988779720835]
xAIL_0_0 : [-3.14159, 3.141590000000001] = [-0, 0]
xAIL_0_t : [-3.14159, 3.141590000000001] = [0.1, 0.1000000000000001]
xAIL_1_0 : [-3.14159, 3.141590000000001] = [0.1, 0.1000000000000001]
xAIL_1_t : [-3.14159, 3.141590000000001] = [0.2, 0.2000000000000001]
xAIL_2_0 : [-3.14159, 3.141590000000001] = [0.2, 0.2000000000000001]
xAIL_2_t : [-3.14159, 3.141590000000001] = [0.2999999999999999, 0.3000000000000001]
xAIL_3_0 : [-3.14159, 3.141590000000001] = [0.2999999999999999, 0.3000000000000001]
xAIL_3_t : [-3.14159, 3.141590000000001] = [0.3999999999999999, 0.4000000000000002]
xAIL_4_0 : [-3.14159, 3.141590000000001] = [0.3999999999999999, 0.4000000000000002]
xAIL_4_t : [-3.14159, 3.141590000000001] = [0.4999999999999998, 0.5000000000000003]
xAIL_5_0 : [-3.14159, 3.141590000000001] = [0.4999999999999998, 0.5000000000000003]
xAIL_5_t : [-3.14159, 3.141590000000001] = [0.4999999999999998, 0.5000696958803174]
xRDR_0_0 : [-3.14159, 3.141590000000001] = [-0, 0]
xRDR_0_t : [-3.14159, 3.141590000000001] = [0, 0]
xRDR_1_0 : [-3.14159, 3.141590000000001] = [-0, 0]
xRDR_1_t : [-3.14159, 3.141590000000001] = [0, 0]
xRDR_2_0 : [-3.14159, 3.141590000000001] = [-0, 0]
xRDR_2_t : [-3.14159, 3.141590000000001] = [0, 0]
xRDR_3_0 : [-3.14159, 3.141590000000001] = [-0, 0]
xRDR_3_t : [-3.14159, 3.141590000000001] = [0, 0]
xRDR_4_0 : [-3.14159, 3.141590000000001] = [-0, 0]
xRDR_4_t : [-3.14159, 3.141590000000001] = [0, 0]
xRDR_5_0 : [-3.14159, 3.141590000000001] = [-0, 0]
xRDR_5_t : [-3.14159, 3.141590000000001] = [0, 0]
delta-sat with delta = 0.00100000000000000
For k = 5, dReach tried 1 feasible paths, all of them are unsat.
```

In the SMT formula, the flow is defined by:

```
(define-ode flow_1 
((= d/dt[beta] (+ (+ (- (* -0.099593 beta) r) (* (/ 9.80555 92.8277) phi)) (* 0.740361 xRDR))) 
(= d/dt[p] (+ (+ (+ (+ (* -1.70098 beta) (* -1.18465 p)) (* 0.223908 r)) (* 0.531304 xAIL)) (* 0.049766 xRDR))) 
(= d/dt[phi] p) 
(= d/dt[psi] (* (/ 9.80555 92.8277) phi)) 
(= d/dt[r] (+ (+ (+ (+ (* 0.40742 beta) (* -0.056276 p)) (* -0.18801 r)) (* 0.005685 xAIL)) (* -0.106592 xRDR))) 
(= d/dt[rAIL] 0) 
(= d/dt[rRDR] 0) 
(= d/dt[tau] 1) 
(= d/dt[xAIL] rAIL) 
(= d/dt[xRDR] rRDR)))
```

but the integral terms have the form with a different variable ordering:

```
(= [beta_1_t p_1_t phi_1_t psi_1_t rAIL_1_t rRDR_1_t r_1_t tau_1_t xAIL_1_t xRDR_1_t] (integral 0. time_1 [beta_1_0 p_1_0 phi_1_0 psi_1_0 rAIL_1_0 rRDR_1_0 r_1_0 tau_1_0 xAIL_1_0 xRDR_1_0] flow_1)
```

Also, there is a problem in the visualization, where the variables r and rRDR do not appear:
![screen shot 2015-04-26 at 9 15 01 pm](https://cloud.githubusercontent.com/assets/8316590/7340334/554af4f8-ec59-11e4-889c-07b326a0bfa9.png)

**Comment by soonhokong** [comment]
I will take a look at this one this morning.

>  I guess that it is cause by different variable orderings of define-ODE and integral.

dReach in dReal3 was changed as well to work with the new variable ordering policy that we have in dReal3. If you have `.smt2` file generated from dReach in dReal2, highly likely it will not work with dReach in dReal3.

> Also, there is a problem in the visualization, where the variables r and rRDR do not appear:

When a variable is constant in a flow, that is `d/dt[variable] = 0`, we won't show it in the visualization. That explain why rAIL and rRDR do not show up in the visualization. I have no idea why r is not there. I'll check that.

**Comment by kquine** [comment]
I DID use dReach in dReal3. FYI, the bug seems to be sensitive to variable
names, particularly
those beginning with the letter 'r'. I have changed every r-variable to
another name, and then
everything works correctly.


On Mon, Apr 27, 2015 at 9:07 AM, Soonho Kong <notifications@github.com>
wrote:

> I will take a look at this one this morning.
>
> I guess that it is cause by different variable orderings of define-ODE and
> integral.
>
> dReach in dReal3 was changed as well to work with the new variable
> ordering policy that we have in dReal3. If you have .smt2 file generated
> from dReach in dReal2, highly likely it will not work with dReach in dReal3.
>
> Also, there is a problem in the visualization, where the variables r and
> rRDR do not appear:
>
> When a variable is constant in a flow, that is d/dt[variable] = 0, we
> won't show it in the visualization. That explain why rAIL and rRDR do not
> show up in the visualization. I have no idea why r is not there. I'll check
> that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dreal/dreal3/issues/73#issuecomment-96643523>.
>



-- 
Kyungmin Bae, Postdoctoral Fellow
School of Computer Science
Carnegie Mellon University

**Comment by soonhokong** [comment]
@kquine, you're right. I'm fixing the problem.

**Comment by soonhokong** [comment]
Fixed by 4d791207e02bc86ddae53272db86d7212f2141f4. The problem was that we sorted the ode variables after adding the `_0`, `_t` suffixes, which can change the ordering.

@kquine, please check out the latest version and recompile dReach by running `make` at `dreal/tools` directory.
